json/json.go

```diff
@@ -305,10 +305,13 @@ func (dec *Decoder) readValue() (v []byte, err error) {
 			dec.buffer = buf
 		}
 
-		n, err = dec.reader.Read(dec.buffer[len(dec.buffer):cap(dec.buffer)])
+		n, err = io.ReadFull(dec.reader, dec.buffer[len(dec.buffer):cap(dec.buffer)])
```

Comment:
Did we evaluate using a bytes.Buffer for dec.buffer? Then we could just do `dec.buffer.ReadFrom(dec.reader))`, and we wouldn't have to handle growing the buffer, as we do above. The reason it's on my mind is that I thought using ReadFull might be a code smell here -- it's designed for the case where we expect at least len(buf) bytes to be available, and we don't expect that here, which is why we have to catch the ErrUnenxpectedEOF.

Just wondering if the code here would be tidier that way, and perhaps it doesn't make a difference performance-wise.

Reply:
I don't think `bytes.Buffer` would work, the only method available to read from an `io.Reader` to a `bytes.Buffer` is `ReadFrom`, and that would load the data in memory until the `io.EOF` is returned. The purpose of the `json.Decoder` is to support reading streams of values, so we want to avoid loading the whole content in memory if possible.

Reply:
I don't follow -- you're using ReadFull here, which does the same thing that ReadFrom would in this case, except that it wouldn't consider EOF an error (which ReadFull does).

Reply:
It doesn't, I'm using `io.ReadFull` to fill the input buffer, `bytes.(*Buffer).ReadFrom` will grow the input buffer to hold all data until io.EOF.

Reply:
🤦 got it. My brain glossed over this difference because I saw the buffer growing code above. Thanks for the explanation.